oc new-app: propagate cmdline env to pipeline/git builds

wozniakjan opened the pull request:
The `oc new-app --build-env` cmdline env variables are propagated to the `BuildConfig` only when using a template. With git repository as a source, env variables can be silently dropped if it contains no template.

This fixes  #19142 ~~by joining the env for `objects` from component references similarly like it is done for `templateObjects` from template component refs in `(c *AppConfig) buildTemplates`.~~

~~Looking at the `BuildConfig` with this fix, env variables are propagated, however, I am not sure how it obtains the Jenkinsfile because neither embedded Jenkinsfile nor path to it contain anything.~~

```json
"Strategy": {
    "DockerStrategy": null,
    "SourceStrategy": null,
    "CustomStrategy": null,
    "JenkinsPipelineStrategy": {
        "JenkinsfilePath": "", 
        "Jenkinsfile": "", 
        "Env": [
           {
                "Name": "PROJECT_NAME",
                "Value": "dev",
                "ValueFrom": null
            }
        ]
    }
},
```

ptal @openshift/sig-developer-experience ~~, I am not sure if this is a valid approach. In the meanwhile, I will try to work on an extended test covering this.~~

bparees commented:
/assign @gabemontero

bparees commented on pkg/oc/generate/cmd/newapp.go:
```diff
@@ -859,6 +869,23 @@ func (c *AppConfig) Run() (*AppResult, error) {
+			if bc, ok := obj.(*buildapi.BuildConfig); ok {
+				buildEnv := buildutil.GetBuildConfigEnv(bc)
+				buildEnv = app.JoinEnvironment(buildEnv, buildenv.List())
+				buildutil.SetBuildConfigEnv(bc, buildEnv)
```
i'm not convinced this is the right place to be changing. 

--build-env already works today for non-pipeline buildconfigs and the codepath you're touching here would impact all buildconfigs i believe.

Something is unique about what new-app does for the "build strategy pipeline" vs what it does for "build strategy s2i/docker/etc" that is causing it to drop the --build-env env vars.

(at a minimum if this logic is added here, we need to find and remove the existing logic that is adding the env vars to all the other buildconfigs)

wozniakjan replied:
thanks for the feedback, I will take a look at other strategies as well

gabemontero replied:
Yeah, to @bparees 's point take a look at https://github.com/openshift/origin/blob/master/pkg/oc/generate/app/app.go#L198-L233

Notice that the env's are set for source/docker build strategy, but not pipeline

Also, you may notice custom strategy is missing there but that is OK ... new-app doesn't deal with that strategy if memory serves me correctly

wozniakjan commented on pkg/oc/generate/cmd/newapp.go:
```diff
@@ -859,6 +869,23 @@ func (c *AppConfig) Run() (*AppResult, error) {
+			}
+			podSpec, _, err := ometa.GetPodSpec(obj)
+			if err == nil {
+				for i := range podSpec.Containers {
```
that may very likely be the case, my intention here was to outline a path I would like to explore. The code is taken from:

https://github.com/openshift/origin/blob/1cfd0d78d24769213faef19a27ffa4a7904d63f2/pkg/oc/generate/cmd/newapp.go#L542-L558

wozniakjan commented:
/retest

miminar commented on test/extended/builds/pipeline.go:
```diff
@@ -776,6 +780,62 @@ var _ = g.Describe("[Feature:Builds][Slow] openshift pipeline build", func() {
 
+			g.By("Pipeline with env vars and git repo source")
+
+			g.By("should should propagate env vars to bc", func() {
```
`s/By/It/`  and drop the second `should`

wozniakjan replied:
looks like I may need to follow the pattern of other tests here and use `By` instead of `It`, getting `You may only call It from within a Describe or Context` error

miminar commented on test/extended/testdata/builds/test-build-app-pipeline-upstream.git/objects/8f/cda77484140980cb3f580797b8efb86b7e2a70:
```diff
@@ -0,0 +1,2 @@
+x��K
+�0@]�s����Ā�W�3�E�S����\�����)8L�"��ыB]���K�D�MR��S�1�3��mL�����[�Sx+z�����<��G�6DX0 ��?'T��3moj�h�FK
```
Can the repository be fetched from remote upon test's initialization instead of making it a kind of submodule?

wozniakjan replied:
yes, it can. The previous commit in this PR is done that way but the available repo is unnecessarily large (as is the case for good number of extended tests and partially a reason, why they take so long) and I thought this might be a good way to speed up the tests - have minimal repos locally.

miminar replied:
I see the benefit although a readable source code is still preferrable over the binary blobs. In case we don't need to perform any remote calls as a part of the test, would something along the following pseudo-bash lines work?

```sh
mkdir testrepo && cd testrepo
touch sourcefile
git init .
git add *
git commit --autor='testrobot <do@not.blame.me>'  -m "nothing to see here"
# use it
```

wozniakjan replied:
I like your idea a lot, thanks! I am pretty sure `github.com/openshift/origin/pkg/git` package will allow this, I will explore it a bit more

bparees commented on pkg/oc/generate/cmd/newapp.go:
```diff
@@ -518,7 +522,13 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
+	environment app.Environment,
+	buildEnvironment app.Environment,
+	templateProcessor templateinternalclient.TemplateProcessorInterface,
+) (string, []runtime.Object, error) {
```
please try to avoid inadvertent formatting changes in PRs (especially when it's the only change in the file).

bparees commented on test/extended/builds/pipeline.go:
```diff
@@ -776,6 +778,63 @@ var _ = g.Describe("[Feature:Builds][Slow] openshift pipeline build", func() {
+				o.Expect(envs[0].Name).To(o.Equal("FOO1"))
+				o.Expect(envs[0].Value).To(o.Equal("BAR1"))
+
+				//TODO: start the build and validate envs as well
```
this shouldn't be necessary, we ought to already have a test somewhere that covers "pipeline buildconfig w/ env vars" behavior.  @gabemontero should be able to point to it.

I think it should be sufficient to confirm the buildconfig that is generated, has the env vars.

wozniakjan replied:
yes, we do, I located this test right below the pipeline buildconfig but I thought since the issue complained about both parts, we may want to test it all through here as well.

coreydaley commented on test/extended/util/framework.go:
```diff
@@ -1310,6 +1311,52 @@ func (r *podExecutor) Exec(script string) (string, error) {
+}
+
+//TODO: add comment
+func (r GitRepo) Cleanup() {
```
Maybe Remove instead of Cleanup since it's just removing a folder? (or two)

coreydaley commented on test/extended/util/framework.go:
```diff
@@ -1310,6 +1311,52 @@ func (r *podExecutor) Exec(script string) (string, error) {
 	return out, waitErr
 }
 
+type GitRepo struct {
```
Also, this struct is really referring to two repositories not one, so the name is slightly misleading (unless you change it to just be the one repo as i suggested below and remove the upstream, then it makes more sense)

wozniakjan replied:
> Also, this struct is really referring to two repositories not one

I took `newapp_test.go` as a reference, and there is `setupLocalGitRepo()` that also creates two repos (and spins up an HTTP server), but is named as if it sets up only one.

https://github.com/openshift/origin/blob/2785d750df87faa4d8cba70afb5f92fb67b73d38/test/integration/newapp_test.go#L2021

I am not really sure what would be better-suited name, all I could come up with is just making it a plural `GitRepos`

> How about using filepath for the upstreamPath and repoPath?

Absolutely, thanks! I also put it in temp dir similarly to `setupLocalGitRepo()`

https://github.com/openshift/origin/blob/2785d750df87faa4d8cba70afb5f92fb67b73d38/test/integration/newapp_test.go#L2023

coreydaley commented:
Hmm, on further inspection, i feel like everything that you added for the GitRepo struct and associated functions are available in the origin/pkg/git package and that this creates semi-duplicate-ish code?
https://github.com/openshift/origin/blob/master/pkg/git/repository.go#L24-L46

wozniakjan commented on test/extended/util/framework.go:
```diff
@@ -1310,6 +1311,52 @@ func (r *podExecutor) Exec(script string) (string, error) {
+}
+
+//TODO: add comment
+func NewGitRepo(repoPath string) (GitRepo, error) {
```
Unfortunately, I think we need upstream repo as well and it was required from `oc new-app` itself.
I found two ways providing source in git repo:
1) URL - and I would like to try to avoid this due to reasons from https://github.com/openshift/origin/pull/19332#discussion_r181662278
2) local path - in this case, it needs to be git repo with upstream (I will try to find the code or at least link to docs, I think I saw it there as well)

coreydaley replied:
I think that, using local path, the repo has to have an accessible origin, not an upstream.

coreydaley replied:
and I am thinking that if it doesn't it will tar up the contents and upload it?  @bparees might know better.

bparees replied:
it needs to have a valid remote. i'm not sure what rules we apply beyond that to pick *which* remote if there are several available.

if you don't have a valid remote, you get a binary build, which in this case might actually be fine.

bparees replied:
again, if we simplify this test to not actually run the build, this probably gets easier.

bparees replied:
yeah i guess that makes sense since you can't do a binary build w/ a pipeline buildconfig.

wozniakjan commented on test/extended/util/framework.go:
```diff
@@ -1310,6 +1311,52 @@ func (r *podExecutor) Exec(script string) (string, error) {
+
+//TODO: add comment
+func NewGitRepo(repoPath string) (GitRepo, error) {
+	upstreamPath := repoPath + `-upstream.git`
```
I will add the upstream repo as second param, seems clearer and less error-prone

wozniakjan commented:
> i feel like everything that you added for the GitRepo struct and associated functions are available in the origin/pkg/git package and that this creates semi-duplicate-ish code?

@coreydaley I agree that at this point it may not look very useful but I would like to reuse the logic of creating minimal git repos for other slow tests https://github.com/openshift/origin/issues/19077. Lot of our extended tests taking ~4 minutes spend 1-2 minutes building an image and 1-2 minutes pushing the image. With minimal repo, I think I could reduce this to a fraction of time. The `origin/pkg/git` package is rather verbose and git test repos will be all very similar so I would like GitRepo to encapsulate same code patterns

wozniakjan commented:
@coreydaley, @bparees, @gabemontero  I think I have addressed all review comments except naming `GitRepo`, I am still unable to come up with anything better, and the integration test is also fixed

bparees commented:
/lgtm

gabemontero commented:
yep the env var injection is in the right spot now

/lgtm as well

wozniakjan commented:
/test cmd

flake?
```
FAILURE after 159.520s: hack/lib/start.sh:284: executing 'oc get --raw /healthz --as system:unauthenticated --config='/tmp/openshift/test-cmd/openshift.local.config/master/admin.kubeconfig'' expecting any result and text 'ok'; re-trying every 0.25s until completion or 160.000s: the command timed out
```

wozniakjan commented:
/test extended_builds
